src/shared/firewall-util-nft.c

```diff
@@ -0,0 +1,853 @@
+         * sizeof(uint8_t) + sizeof(uint16_t).
+         */
+        assert_return(msgcnt < NFT_INIT_MSGS, -EINVAL);
+        r = nft_new_map(&ruleset[++msgcnt], nft_systemd_dnat_map_name, 0,
+                        concat_types2(TYPE_INET_PROTOCOL, TYPE_INET_SERVICE), sizeof(uint32_t) * 2,
+                        concat_types2(TYPE_IPADDR, TYPE_INET_SERVICE), sizeof(uint32_t) * 2);
+        if (r < 0)
+                return r;
+
+        assert_return(msgcnt < NFT_INIT_MSGS, -EINVAL);
+        r = sd_nfnl_message_new_dnat_rule_pre(&ruleset[++msgcnt], "prerouting");
+        if (r < 0)
+                return r;
+
+        assert_return(msgcnt < NFT_INIT_MSGS, -EINVAL);
+        r = sd_nfnl_message_new_dnat_rule_out(&ruleset[++msgcnt], "output");
+        if (r < 0)
+                return r;
+
+        assert_return(msgcnt < NFT_INIT_MSGS, -EINVAL);
+        r = sd_nfnl_message_new_masq_rule(&ruleset[++msgcnt], "postrouting");
+        if (r < 0)
+                return r;
+
+        return msgcnt;
+}
+
+int fw_nftables_init(void)
+{
+        _cleanup_(sd_netlink_unrefp) sd_netlink *nfnlp;
+        sd_netlink_message *ruleset[NFT_INIT_MSGS];
+        unsigned int msgcnt = 0;
+        int r;
+
+        if (nfnl)
+                return 0;
+
+        memset(ruleset, 0, sizeof(ruleset));
+
+        r = sd_nfnl_socket_open(&nfnlp);
+        if (r < 0)
+                return r;
+
+        nfnl = sd_netlink_ref(nfnlp);
+        assert_return(nfnlp, -EINVAL);
+
+        r = sd_nfnl_message_batch_begin(nfnl, &ruleset[msgcnt]);
+        if (r < 0)
+                goto out_unref;
+
+        assert_return(msgcnt < NFT_INIT_MSGS, -EINVAL);
+        r= __fw_nftables_init(ruleset, msgcnt);
+        if (r < 0)
+                goto out_unref;
+
+        assert_return(r > (int)msgcnt, -EINVAL);
+
+        msgcnt = r;
+
+        assert_return(msgcnt < NFT_INIT_MSGS, -EINVAL);
+        r = sd_nfnl_message_batch_end(nfnl, &ruleset[++msgcnt]);
+        if (r < 0)
+                goto out_unref;
+
+        r = nfnl_netlink_sendv(nfnl, ruleset, ++msgcnt);
+        if (r < 0)
+                return r == -EEXIST ? 0 : r;
+
+        return 0;
+out_unref:
+        sd_netlink_unref(nfnl);
+        while (msgcnt)
+                sd_netlink_message_unref(ruleset[--msgcnt]);
+        return r;
+}
+
+static int nft_message_add_setelem_iprange(sd_netlink_message *m,
+                                           const union in_addr_union *source,
+                                           unsigned int prefixlen)
+{
+        union in_addr_union source_end = *source;
+        unsigned int nplen = 32 - prefixlen;
+        uint32_t v = (1 << nplen);
+        int r;
+
+        r = sd_nfnl_nft_message_add_setelem(m, 0, &source->in, sizeof(uint32_t), NULL, 0);
+        if (r < 0)
+                return r;
+        r = sd_nfnl_nft_message_add_setelem_end(m);
+        if (r < 0)
+                return r;
+
+        v = (1 << nplen);
+
+        source_end.in.s_addr += htonl(v);
```

Comment:
please do not use htonl() in our codebase. it's really weird, since it suggests it would return "long" but actually returns uint32_t. Also, "network byte order" is just legacy talk for "big endian". Hence, please htobe32() instead, which is much clearer.